ui/components-react/src/components-react/newEditors/editors/numeric/FormattedNumericInput.tsx

```diff
@@ -0,0 +1,120 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import * as React from "react";
+import { Input } from "@itwin/itwinui-react";
+import type { NumericValue } from "../../values/Values.js";
+
+/**
+ * Props for FormattedNumericInput component.
+ * @beta
+ */
+interface FormattedNumericInputProps {
+  value: NumericValue;
+  onChange: (value: NumericValue) => void;
+  parseValue: (value: string) => number | undefined;
+  formatValue: (num: number) => string;
+  onBlur: () => void;
+  disabled?: boolean;
+  size?: "small" | "large";
+}
+
+/**
+ * A numeric input that allows to pass custom parsing/formatting logic to handle values with units etc.
+ * @beta
+ */
+export function FormattedNumericInput({
+  onChange,
+  value,
+  parseValue,
+  formatValue,
+  disabled,
+  onBlur,
+  size,
+}: FormattedNumericInputProps) {
+  const { currentValue, inputProps } = useParsedNumberInput({
+    initialValue: value.rawValue,
+    parseValue,
+    formatValue,
+  });
+  const onChangeRef = React.useRef(onChange);
+  React.useEffect(() => {
+    onChangeRef.current = onChange;
+  }, [onChange]);
+
+  React.useEffect(() => {
+    onChangeRef.current(currentValue);
+  }, [currentValue]);
+
+  return (
+    <Input disabled={disabled} size={size} onBlur={onBlur} {...inputProps} />
+  );
+}
+
+function useParsedNumberInput({
+  initialValue,
+  formatValue,
+  parseValue,
+}: {
+  initialValue: number | undefined;
+  parseValue: (value: string) => number | undefined;
+  formatValue: (num: number) => string;
+}) {
+  interface State {
+    value: NumericValue;
+    placeholder: string;
+  }
+
+  const [state, setState] = React.useState<State>(() => {
+    return {
+      value: {
+        rawValue: initialValue,
+        displayValue:
+          initialValue !== undefined ? formatValue(initialValue) : "",
+      },
+      placeholder: formatValue(123.45),
```

Comment:
maybe use `initialValue` as the placeholder, if it's provided?

Reply:
Switched to `initialValue`